Comprehensive validation 🔎, 30+ fixes integrated/added 🔨🐛, optimized performance 🚀

klondikedragon opened the pull request:
This package is amazing and hugely popular, and has been the best package for automatic date parsing in go for years! ⭐

Thanks @araddon for crafting this package with love over the years!!

I've been using this while developing a new cloud-based log aggregation/search/visualization product, and I've found that there are three major opportunities for improvement for my particular use case:

* The package does not strictly validate its input, leading to many false positives. This is OK if you know for sure the input matches one of the known formats, but cannot be trusted if the input could be anything and you only want a returned date/time if it definitely matches a known format.
* While still being far more efficient than the "shotgun" parsing approach, the package currently allocates a relatively large amount of memory (several times the average input size), which can add up when parsing megabytes of date strings per second in a high-throughput microservice. It can also be relatively slow when parsing a string that doesn't match a known format and can allocate even more memory in this case, due to custom error messages that include contextual details.
* There are a lot of unmerged one-off contributions that haven't been merged and need to be made coherent with each other.

This PR addresses all 3 opportunities:
* **Validation**: it now comprehensively validates the input in the following ways: at each point in the state machine, it makes sure there are cases for all possibilities, and any invalid possibility will fail; additionally, it makes sure that the entire format string was specifically set (excluding any trailing punctuation, which can be safely ignored). False positives should be extremely rare now (hard to prove they don't exist).
* **Memory Efficiency**: bytes allocated have been reduced by 90%, and parsing some formats are zero allocation. The `SimpleErrorMessages` option was added (off by default) that greatly speeds up the case where a string does not match a known format -- with the option on, this case is now 4x faster and produces almost no allocations.
* **Merge & Integrate Community Fixes**: Fixes for nearly all of the pending issues from the community (including open pull requests) have been incorporated and adapted.

In the process of going through the state machine comprehensively for validation, redundant code/states were merged, and support was added for certain edge cases (for example, some date formats did not support being followed by times).

The example and README.md were updated to incorporate all of the newly supported formats and edge cases. More details on how to properly interpret returned location information with respect to abbreviated timezones was added.

**BREAKING** -- the package now requires go >= 1.20 to support memory optimizations converting from `[]byte` to `string` in key places.

A huge thanks to all who posted issues and contributed PRs -- while the PRs were unable to be merged directly because the validation changes were so major, the ideas of all these contributions and the associated test cases were incorporated. Here's credit for all of the issues fixes and contributions in this PR as well as a summary of additional fixes added:
* This PR started directly building from https://github.com/araddon/dateparse/pull/151 by @arran4 that fixes https://github.com/araddon/dateparse/issues/78 (support for GMT+offset, support additional whitespace in certain places, and other changes)
* Fix https://github.com/araddon/dateparse/issues/145 - words falsely detected as a valid format - fixed by comprehensive validation
* Fix https://github.com/araddon/dateparse/issues/108 - `"1.jpg"` incorrectly parsed as valid - fixed by comprehensive validation
* Fix https://github.com/araddon/dateparse/issues/98 - names and addresses incorrectly parsed as valid - fixed by comprehensive validation
* Fix https://github.com/araddon/dateparse/issues/150 - fix parsing "am" as "pm"
* Fix https://github.com/araddon/dateparse/issues/157 - fix parsing of `Thu Jan 28 2021 15:28:21 GMT+0000` format
* Fix https://github.com/araddon/dateparse/issues/137 - variant of ':' as separator for fraction seconds - adapted https://github.com/araddon/dateparse/pull/138 by @dferstay
* Fix https://github.com/araddon/dateparse/issues/139 - add support for dd-mm-yyyy (digit month) - adapt https://github.com/araddon/dateparse/pull/140 by @dferstay (also fixes https://github.com/araddon/dateparse/issues/155)
* Fix https://github.com/araddon/dateparse/issues/141 - add support for `yyyy mon dd` format - adapt https://github.com/araddon/dateparse/pull/142 by @dferstay
* Fix https://github.com/araddon/dateparse/issues/143 - add support for `yyyymmddhhmmss.SSS` format - adapt https://github.com/araddon/dateparse/pull/144 by @dferstay
* Fix https://github.com/araddon/dateparse/issues/130 - fix timezone parsing after fractional seconds - adapted https://github.com/araddon/dateparse/pull/131 by @zifengyu 
* Fix https://github.com/araddon/dateparse/issues/123 - fix parsing of fractional seconds for certain formats (e.g., `2017-04-03 22:32:14.322 CET`)
* Fix https://github.com/araddon/dateparse/issues/109 - fix parsing of `Sun, 07 Jun 2020 00:00:00 +0100` format
* Fix bug in comment https://github.com/araddon/dateparse/issues/100#issuecomment-1118868154 - fix parsing of `1 April 2022 23:59` format (time after certain date formats)
* Fix https://github.com/araddon/dateparse/issues/129 - fix ambiguous and PreferMonthFirst parsing for format `mm.dd.yyyy (time)` - adapted https://github.com/araddon/dateparse/pull/133 by @mehanizm (also fixes https://github.com/araddon/dateparse/issues/91 and https://github.com/araddon/dateparse/issues/28)
* Fix https://github.com/araddon/dateparse/issues/149 - support `PMDT` and `AMT` time zones and validate that AM/PM indicators only appear at most once
* Fix https://github.com/araddon/dateparse/issues/127 - add support for `dd[th,nd,st,rd] Month yyyy` format - adapt https://github.com/araddon/dateparse/pull/128 by @krhubert
* Fix https://github.com/araddon/dateparse/issues/158 - fix parsing for format `(time) UTC[+-]NNNN`
* Add support for `mm/dd/yyyy, hh:mm:ss` format - adapt https://github.com/araddon/dateparse/pull/156 by @BrianLeishman
* Add support for `yyyy.mm.dd (time)` format - adapt https://github.com/araddon/dateparse/pull/134 by @jmdacruz, and add cases expected to fail to `TestParseErrors` unit test
* Add support for git log format (e.g., `Thu Apr 7 15:13:13 2005 -0700`) - adapt commit https://github.com/araddon/dateparse/pull/92/commits/99d9682a1cbe7a14975b5b71704af8847c2684f9 from https://github.com/araddon/dateparse/pull/92 by @jiangxin (merge timeWsYearOffset case and add validation)
* Add support for RabbitMQ log format (e.g., `dd-mon-yyyy::hh:mm:ss`) - adapt https://github.com/araddon/dateparse/pull/122 by @bizy01
* Expand support for Chinese date formats, and allow times to follow - adapt https://github.com/araddon/dateparse/pull/132 by @xwjdsh
* Add support for `mon/dd/yyyy` format, e.g., `Oct/31/1970`
* Add support for `dd-month-year` format
* Extend format `yyyy-mon-dd` to allow times to follow it. Also allow full month name instead of just abbreviated.
* Fix the case for ambiguous date/time in the `mm:dd:yyyy` format
* Allow full day name before month (e.g., `Monday January 4th, 2017`)
* Additional fixes for `mm.dd.yyyy (time)` format
* Fix ambiguous parsing for `mm/dd` formats that start with a weekday

Also adds tests to verify that the following stay fixed:
* https://github.com/araddon/dateparse/issues/94 - mysql log format such as `190910 11:51:49`
* https://github.com/araddon/dateparse/issues/117 - fractional seconds after ':'

arran4 committed:
added some actions

arran4 committed:
My issue

arran4 committed:
Error return value is not checked (errcheck)

arran4 committed:
field `offsetlen` is unused (unused)

arran4 committed:
S1021: should merge variable declaration with assignment on next line (gosimple)

arran4 committed:
S1023: redundant break statement (gosimple)

arran4 committed:
SA4006: this value of `err` is never used (staticcheck)

arran4 committed:
Lint action out of date.

arran4 committed:
Go mod tidy

arran4 committed:
Bug fixes.

arran4 committed:
More typo changes

arran4 committed:
Another error

arran4 committed:
S1023: redundant break statement (gosimple)

arran4 committed:
Text should be lowercase

arran4 committed:
Added go releaser

arran4 committed:
Commented code

arran4 committed:
Unnecessary bracket

arran4 committed:
Test improvements.. I think

arran4 committed:
My addition last

arran4 committed:
So people don't have to check the string they can use the new errors.Is function

arran4 committed:
Skip white space - to delete strategically

arran4 committed:
All of these did nothing

arran4 committed:
The only required one.

arran4 committed:
New failure - still white space

arran4 committed:
Skip white space

arran4 committed:
Unused code

arran4 committed:
Another case.

klondikedragon committed:
Fix ineffective break statements

klondikedragon committed:
Add extensive format validation, bugfixes

* Don't just assume we were given one of the valid formats.
* Also consolidate the parsing states that occur after timePeriod.
* Add subtests to make it easier to see what fails.
* Additional tests for 4-char timezone names.
* Fix https://github.com/araddon/dateparse/issues/117
* Fix https://github.com/araddon/dateparse/issues/150
* Fix https://github.com/araddon/dateparse/issues/157
* Fix https://github.com/araddon/dateparse/issues/145
* Fix https://github.com/araddon/dateparse/issues/108
* Fix https://github.com/araddon/dateparse/issues/137
* Fix https://github.com/araddon/dateparse/issues/130
* Fix https://github.com/araddon/dateparse/issues/123
* Fix https://github.com/araddon/dateparse/issues/109
* Fix https://github.com/araddon/dateparse/issues/98
* Addresses bug in https://github.com/araddon/dateparse/issues/100#issuecomment-1118868154

Adds test cases to verify the following are already fixed:
* https://github.com/araddon/dateparse/issues/94

klondikedragon committed:
Incorporate fix for dd.mm.yyyy format

Incorporates PR https://github.com/araddon/dateparse/pull/133 from https://github.com/mehanizm to fix https://github.com/araddon/dateparse/issues/129

Adds test cases to verify the following are already fixed:
* https://github.com/araddon/dateparse/issues/105

klondikedragon committed:
Support PMDT and AMT time zones

Also disallow PM and AM from being specified twice in the string.

Fixes https://github.com/araddon/dateparse/issues/149

klondikedragon committed:
Add support for dd[th,nd,st,rd] Month yyyy

Incorporate PR https://github.com/araddon/dateparse/pull/128 from
https://github.com/krhubert to fix
https://github.com/araddon/dateparse/issues/127

klondikedragon committed:
Add support for mon/dd/yyyy (Oct/31/1970)

klondikedragon committed:
Implement support for yyyy mon dd (2013 May 02)

Incorporate PR https://github.com/araddon/dateparse/pull/142 from
https://github.com/dferstay to fix
https://github.com/araddon/dateparse/issues/141

klondikedragon committed:
Incorporate support for dd-mm-yyyy (digit month)

Incorporate PR https://github.com/araddon/dateparse/pull/140 from
https://github.com/dferstay to fix
https://github.com/araddon/dateparse/issues/139

This also fixes https://github.com/araddon/dateparse/issues/155
(duplicate of issue 139)

PR is adapted to avoid duplicate code and validate format.

klondikedragon committed:
Incorporate support for yyyymmddhhmmss.SSS

Incorporate PR https://github.com/araddon/dateparse/pull/144 from
https://github.com/dferstay to fix
https://github.com/araddon/dateparse/issues/143

klondikedragon committed:
Add tests to verify ambiguous cases

Test cases now validates the following is true:
* Fixed https://github.com/araddon/dateparse/issues/91
* Fixed https://github.com/araddon/dateparse/issues/28

(previous commits already addresses these issues, these tests ensure
that these issues remain fixed)

klondikedragon committed:
Handle format "date time (MST)"

Was unable to handle standalone timezone in parentheses before.

Also update tests to indicate expected timezone name for all tests that are parsed in a specific location.

With updated logic/fixes, add tests to verify:
* Fix https://github.com/araddon/dateparse/issues/71
* Fix https://github.com/araddon/dateparse/issues/72

klondikedragon committed:
Add better timezone explanation to README.md

How golang parses date strings with respect to time zones and locations can be really confusing. Document the key points that need to be understood to properly interpret the results of parsing arbitrary date strings, which may or may not have explicit time zone name or offset information include the parsed date string.

klondikedragon committed:
Fix parsing for format (time) UTC[+-]NNNN

Fixes https://github.com/araddon/dateparse/issues/158

klondikedragon committed:
Add support for mm/dd/yyyy, hh:mm:ss

Incorporate PR https://github.com/araddon/dateparse/pull/156 from https://github.com/BrianLeishman and adapt to also validate the format

klondikedragon committed:
Support times after yyyy.mm.dd dates

Fix for this bug mentioned in https://github.com/araddon/dateparse/pull/134

Also, the other cases mentioned in this PR are not valid formats, so add them to the TestParseErrors test, to document that this is expected.

klondikedragon committed:
Expand Chinese date format support

Inspired by https://github.com/araddon/dateparse/pull/132 from https://github.com/xwjdsh -- made this more general to all time formats that could follow, and added format validation.

Also include the related README.md touchup from https://github.com/araddon/dateparse/pull/136

klondikedragon committed:
Support git log format (Thu Apr 7 15:13:13 2005 -0700)

Adapt commit https://github.com/araddon/dateparse/pull/92/commits/99d9682a1cbe7a14975b5b71704af8847c2684f9 from https://github.com/araddon/dateparse/pull/92 by https://github.com/jiangxin (merge timeWsYearOffset case and validate format)

klondikedragon committed:
Support RabbitMQ log format (dd-mon-yyyy::hh:mm:ss)

Adapt https://github.com/araddon/dateparse/pull/122 by https://github.com/bizy01 to add support for RMQ log format. Refactor to avoid redundant code. Add format validations.

As a side note, will also support the format dd-mm-yyyy:hh:mm:ss.

klondikedragon committed:
Add broader benchmarks

Uses the main test set for a broader stress test.

klondikedragon committed:
Heavily optimize memory allocations

Uses a memory pool for parser struct and format []byte

Uses a new go 1.20 feature to avoid allocations for []byte to string conversions in allowable cases.

go 1.20 also fixes a go bug for parsing fractional sec after a comma, so we can eliminate a workaround.

The remaining allocations are mostly unavoidable (e.g., time.Parse constructing a FixedZone location or part to strings.ToLower).

Results show an 89% reduction in allocated bytes for the big benchmark cases, and for some formats an allocation can be avoided entirely.

There is also a resulting 26% speedup in ns/op.

Details:

BEFORE:

cpu: 12th Gen Intel(R) Core(TM) i7-1255U
BenchmarkShotgunParse-12               19448 B/op        474 allocs/op
BenchmarkParseAny-12                    4736 B/op         42 allocs/op
BenchmarkBigShotgunParse-12          1075049 B/op      24106 allocs/op
BenchmarkBigParseAny-12               241422 B/op       2916 allocs/op
BenchmarkBigParseIn-12                244195 B/op       2984 allocs/op
BenchmarkBigParseRetryAmbiguous-12    260751 B/op       3715 allocs/op
BenchmarkShotgunParseErrors-12         67080 B/op       1679 allocs/op
BenchmarkParseAnyErrors-12             15903 B/op        200 allocs/op

AFTER:

BenchmarkShotgunParse-12               19448 B/op        474 allocs/op
BenchmarkParseAny-12                      48 B/op          2 allocs/op
BenchmarkBigShotgunParse-12          1075049 B/op      24106 allocs/op
BenchmarkBigParseAny-12                25394 B/op        824 allocs/op
BenchmarkBigParseIn-12                 28165 B/op        892 allocs/op
BenchmarkBigParseRetryAmbiguous-12     37880 B/op       1502 allocs/op
BenchmarkShotgunParseErrors-12         67080 B/op       1679 allocs/op
BenchmarkParseAnyErrors-12              3851 B/op        117 allocs/op

klondikedragon committed:
Optimize ambiguous date parsing

Previously, for ambiguous date strings, it was always calling parse twice even when the first parse would have been successful.

Refactor so that parsing isn't re-attempted unless the first parse fails ambiguously.

Benchmark results show that with RetryAmbiguousDateWithSwap(true), it's now about 6.5% faster (ns/op) and reduces allocated bytes by 3.4%.

klondikedragon committed:
Further optimize ambiguous parsing

Optimize the common and special case where mm and dd are the same length, just swap in place. Avoids having to reparse the entire string.

For this case, it's about 30% faster and reduces allocations by about 15%.

This format is especially common, hence the reason to optimize for this case.

Also fix the case for ambiguous date/time in the mm:dd:yyyy format.

klondikedragon committed:
Comprehensive date format validation

Audit every stateDate so every unexpected alternative will fail.

In the process, fixed some newly found bugs:
* Extend format yyyy-mon-dd to allow times to follow it. Also allow full month name.
* Allow full day name before month (e.g., Monday January 4th, 2017)

Relevant confirmatory test cases were added.

klondikedragon committed:
Optimize memory for error case

New option SimpleErrorMessages that avoids allocation in the error path. It's off by default to preserve backwards compatibility.

Added benchmark BenchmarkBigParseAnyErrors that takes the big set of test cases, and injects errors to make them fail at pseudo-random places.

This optimization speeds up the error path runtime by 4x and reduces error path allocation bytes by 13x!

klondikedragon committed:
Optimize checks for day of week and full month

Reduces CPU usage on large benchmarks by ~2%-3% and prepares for future with international month names in future.

klondikedragon committed:
Comprehensive time validation

klondikedragon committed:
Fix mm.dd.yyyy (time) format

klondikedragon committed:
Add support for dd-month-year format

klondikedragon committed:
Update example and README.md with new formats

Audited all test cases to make sure an example was listed for all known formats.

klondikedragon committed:
Fix ambiguous mm/dd that start with weekday

Options were not being properly passed to recursive parseTime call.

klondikedragon committed:
Update benchmark results

klondikedragon committed:
Update go doc

arran4 commented:
Great work @klondikedragon

jmdacruz commented:
this is great work @klondikedragon! Now, this repo hasn't seen much movement in years, do you think we should start using a fork? should we use yours?

arran4 commented:
I would vote that we use his, we should see if it qualifies for https://github.com/avelino/awesome-go

klondikedragon committed:
Allow weekday prefix for most date formats

This is implemented now using the "skip" parser field, indicating
to skip the first N characters. This also avoids a recursive parse
in one more case (more efficient). This simplifies the state machine
a little bit, while the rest of the code needs to properly account
for the value of the skip field.

Also allow whitespace prefix without penalty.

Modify the test suite to psuedo-randomly add a weekday prefix
to the formats that allow it (all except the purely numeric ones).

klondikedragon commented:
In some further testing, I found that weekday prefixes only worked for some date formats, but not for others. So that is fixed now. As a side effect (benefit?), leading whitespace is now allowed/ignored.

Let's see if @araddon has feedback and/or is interested in merging this PR (it's a pretty big change and changes the philosophy a bit to have validation, and also makes the code a little more complex in favor of performance). The changes are large enough now it could break backwards compatibility, so in the very least it should deserve a new major version IMO.

Although I don't want to fork something lightly, since we haven't heard any feedback from @araddon for a few years, it could definitely make sense to go ahead. If there is no comment after the holidays, I think it would make sense to go ahead and fork. This package is a key part of freeform date parsing in the "automatic structured field extraction" logic being built for IT Lightning (this new cloud-based log management platform I'm building). Given that's the case, the IT Lightning org would be willing to maintain the forked repo and work with community issues/contributions, since we're motivated to have best-in-class date recognition & parsing in the log ingestion pipeline. The license would remain the same of course.

The community contributions would help us improve our date parsing, we'd be motivated to put energy into it to keep our date parsing bug-free and comprehensive, and community use of the package might help us get a little exposure to devs/SREs who might become interested in our log management solution. So it should benefit everyone.

All feedback is welcome. What do ya'll think of this proposed plan?

klondikedragon committed:
Unify/fix timezone offset/name states

* Merge duplicate states (fixes lots of edge cases)
* Support for +00:00 is consistent with +0000 now
* Support (timezone description) after any offset/name
* Update tests to cover positive/negative cases
* Update example with new supported formats

klondikedragon committed:
Cleanup handling of TZ name parsing

Fully support the format where a TZ name is in parentheses after the
time (and possibly after an offset). This fixes the broken case where a
4 character TZ name was in parentheses after a time.

elliot40404 commented:
great work @klondikedragon . How can i start using this?

klondikedragon commented:
I'll go ahead and fork this package. I'm renaming the main branch as part of that.

klondikedragon committed:
Fork package to github.com/itlightning/dateparse

Various other cleanup:
* Update README.md
* Update github workflows
* Add to copyright
* Add .gitignore

klondikedragon committed:
Merge pull request #1 from itlightning/prerelease/v0.1.0

Prepare v0.1.0 release

klondikedragon committed:
Fix goreleaser github action

klondikedragon commented:
The fork is complete and published as `v0.1.0` -- again, a huge thanks to @araddon for authoring and maintaining this package for so many years!

The fork is available using `go get github.com/itlightning/dateparse` -- issues and PRs are welcome.

@elliot40404 @arran4 @jmdacruz -- see what you think and how this updated package works! If this looks good and after incorporating feedback, I think I'll publish a v1.0.0 at some point soon. I'm also curious to get feedback on my log management project too, check out the site/discord if you're interested. Thanks!

elliot40404 committed:
fixed lib import

klondikedragon committed:
Merge pull request #2 from elliot40404/patch-1

fixed lib import

klondikedragon committed:
(chore) upgrade deps

klondikedragon committed:
Merge pull request #3 from itlightning/chore/2024-02-17-upgrade-deps

(chore) upgrade deps

klondikedragon committed:
Add prefer-day-first flag and relatedto example program

Support --prefer-day-first and --retry-ambiguous flags for example
program. Also provide an example where parsing would fail without it.

klondikedragon committed:
Upgrade github actions

klondikedragon committed:
Upgrade go dependencies to latest

klondikedragon committed:
Fix latest lint errors

klondikedragon committed:
Merge pull request #6 from itlightning/enhance-dateparse-example

Enhance dateparse example with flags related to ambiguous formats

klondikedragon committed:
Add support for RFC3164/RFC5424 formats

Many devices send dates that do not conform to the RFCs...

Also add support for the strange "TZ-0700" variant of the "UTC-0700"
offset.

Cover all the changes with new tests.

klondikedragon committed:
Merge pull request #7 from itlightning/syslog-formats

Support for RFC3164/RFC5424 syslog formats

arran4 commented:
What happened?

klondikedragon commented:
@arran4 sorry... I was cleaning up old branches when releasing a new version of https://github.com/itlightning/dateparse (part of work to add automated config-free syslog parsing/ingestion to https://sparklogs.com/ ), and I forgot that deleting the old `master` branch would close this PR (I'm using `main` branch actively now).  I've recreated the branch so this PR can remain open 🙂 

v0.2.1 of the github.com/itlightning/dateparse package is now available with support for additional syslog RFC3164/RFC5424 time formats (and their many variants). Dependencies were also updated. Use/development of the package continues to be active there and feedback welcome.
